### ISSUE TITLE
Provide a wrapper method that can take long lines to text and work around the 1000 token / 40 second limit

### ISSUE THREAD

**Issue by slidenerd** [resemble-ai/chatterbox#181]
- As you know the library has a major limitation currently
- Give it more than 1000 tokens or anything that takes longer than 40 seconds, it gets cut immediately
- Kindly provide us a wrapper method around this that perhaps chunks into 1000 tokens

**Comment by rsxdalv** [comment]
> * As you know the library has a major limitation currently
> * Give it more than 1000 tokens or anything that takes longer than 40 seconds, it gets cut immediately
> * Kindly provide us a wrapper method around this that perhaps chunks into 1000 tokens

If you chunk it by 1000 tokens, you are going to have artifacts. The solution is to chunk it based on break-points of text, such as paragraphs or sentences. A lot of the projects based on chatterbox have these features.

I think they could add a simple example chunk script that divides based on sentences. Advanced chunking requires bigger and bulkier processing which might be out of scope.

**Comment by ukr-projects** [comment]
Correct, I am working on it and in 4 to 5 days, I will try to contribute.

**Comment by ukr-projects** [comment]
But if you need chunking now check out this repo--> https://github.com/UKR-PROJECTS/chatterbox-tts-colab

**Comment by haydonryan** [comment]
> * As you know the library has a major limitation currently
> * Give it more than 1000 tokens or anything that takes longer than 40 seconds, it gets cut immediately
> * Kindly provide us a wrapper method around this that perhaps chunks into 1000 tokens

You could have just loaded up any model and asked it to code it for you.....

You know like this (splits on paragraphs, doesn't check if there's more than 1000 tokens) but took me 5mins to: go from "write a python app that reads text from a file and calls a “speak” function on each sentence" to the code below.... and I don't know python more than tabs = brackets.  (I also changed the break from sentences to paragraphs since that works better)

Edits: because i'm feeling spicy... this will read a text document and give you a wav file at the end of it
```
import re
import torchaudio as ta
import torch
from chatterbox.tts import ChatterboxTTS

def main():
    # Read the text from the file
    try:
        with open('input.txt', 'r', encoding='utf-8') as file:
            text = file.read()
    except FileNotFoundError:
        print("Error: The file 'input.txt' was not found.")
        return


    AUDIO_PROMPT_PATH = "accusations.wav"
    model = ChatterboxTTS.from_pretrained(device="cuda")
    # Split the text into sentences using regular expressions
    #sentences = re.split(r'[!.?]+', text)
    sentences = re.split(r'\n\n', text)
    filtered_sentences = [s.strip() for s in sentences if s.strip()]
    i=0
    combined_audio=None
    # Call the speak function for each sentence
    for sentence in filtered_sentences:

        wav = model.generate(sentence, audio_prompt_path=AUDIO_PROMPT_PATH)
        filename = "test-" + str(i) + ".wav"

        #if i!=0:
        if combined_audio is None:
            combined_audio= wav # Initialize with the first audio tensor
        else:
            combined_audio = torch.cat((combined_audio,wav),dim=1)

        ta.save(filename, wav, model.sr)
        i+=1

    ta.save("out.wav", combined_audio, model.sr)

if __name__ == "__main__":
    main()

```

**Comment by haydonryan** [comment]
you might need to change the \n\n to just one depending on the input text.

**Comment by haydonryan** [comment]
One curious thing I'm seeing is that the token counts from 
 tok = model.tokenizer.text_to_tokens( paragraph ) and 
 t2 = model.tokenizer.encode(paragraph)

are giving way off counts - about 1/2 of what the actual token counts are when doing the encode.

**Comment by saurabhthesuperhero** [comment]
@haydonryan  its not about using any chunking logic, issue is
for eg i have this chunking code 
`def chunk_text(text: str, max_len: int = 300):
    """Splits text into manageable chunks based on sentences or max length."""
    # Split by sentence-ending punctuation followed by a space
    sentences = re.split(r'(?<=[.!?।])\s+', text)
    chunks = []
    current_chunk = ""
    for sentence in sentences:
        if len(current_chunk) + len(sentence) + 1 <= max_len:
            current_chunk += sentence + " "
        else:
            if current_chunk: # Add the current chunk if it's not empty
                chunks.append(current_chunk.strip())
            current_chunk = sentence + " "
    
    if current_chunk: # Add the last remaining chunk
        chunks.append(current_chunk.strip())
        
    # If there are no sentences, just split by length
    if not chunks:
        chunks = [text[i:i+max_len] for i in range(0, len(text), max_len)]
        
    return chunks`

but it generates noise or speed up or sometime pick up some differnt pace, so recommneded approach will be better.

**Comment by NaumanHSA** [comment]
Easy Solution to tackle long generations, as I'm using and it's working great for me. Set `max_char <= 300 chars`.


    def _split_into_chunks(text: str, max_chars: int) -> list[str]:
        """
        Split *text* into segments no longer than *max_chars*.
    
        Break priority (highest to lowest):
          1. Sentence boundaries  (. ! ?)
          2. Clause boundaries    (, ;)
          3. Word boundaries      (hard split, last resort)
    
        The original punctuation is preserved at the end of each sentence.
        """
        text = text.strip()
        if not text:
            return []
        if len(text) <= max_chars:
            return [text]
    
        # Split into sentences, keeping terminal punctuation attached.
        sentences = re.split(r"(?<=[.!?])\s+", text)
        chunks: list[str] = []
        current = ""
    
        def _flush(buf: str, target: list[str]) -> str:
            if buf:
                target.append(buf)
            return ""
    
        for sent in sentences:
            sent = sent.strip()
            if not sent:
                continue
    
            if len(sent) > max_chars:
                current = _flush(current, chunks)
                # Secondary: clause boundaries
                clauses = re.split(r"(?<=[,;])\s+", sent)
                for clause in clauses:
                    clause = clause.strip()
                    if not clause:
                        continue
                    if len(clause) > max_chars:
                        current = _flush(current, chunks)
                        # Tertiary: word boundaries
                        for word in clause.split():
                            candidate = (current + " " + word).strip() if current else word
                            if len(candidate) <= max_chars:
                                current = candidate
                            else:
                                current = _flush(current, chunks)
                                current = word
                    else:
                        candidate = (current + " " + clause).strip() if current else clause
                        if len(candidate) <= max_chars:
                            current = candidate
                        else:
                            current = _flush(current, chunks)
                            current = clause
            else:
                candidate = (current + " " + sent).strip() if current else sent
                if len(candidate) <= max_chars:
                    current = candidate
                else:
                    current = _flush(current, chunks)
                    current = sent
    
        _flush(current, chunks)
        # Fallback: should never trigger, but guarantee at least one chunk.
        return chunks or [text[:max_chars]]
    
And then, make use of the chunker in actual generations like:

    max_chars_per_chunk = 300
    text_chunks = _split_into_chunks(text, max_chars_per_chunk)
    
    audio_chunks: list = []
      for i, chunk_text in enumerate(text_chunks):
          with self._torch.inference_mode():
              wav = self._model.generate(
                  text=chunk_text,
                  audio_prompt_path=audio_prompt,
                  temperature=self.temperature,
                  top_p=self.top_p,
                  top_k=self.top_k,
                  repetition_penalty=self.repetition_penalty,
                  exaggeration=self.exaggeration,
                  cfg_weight=self.cfg_weight,
                  norm_loudness=self.norm_loudness,
              )

          wav = wav.detach().cpu()
          if wav.dim() == 1:
              wav = wav.unsqueeze(0)
          audio_chunks.append(wav)

          # Release GPU memory between chunks
          gc.collect()
          if self.device == "cuda":
              self._torch.cuda.empty_cache()

      final_wav = self._torch.cat(audio_chunks, dim=-1)
      ta.save(str(output_path), final_wav, self._model.sr)

I hope it will help your requirements.
